### PR TITLE
Added ResUNet and ResUNet++ decoders

### PR DESCRIPTION
**Pull request by jlcsilva**
Implemented the ResUNet and ResUNet++ decoders, as described in https://arxiv.org/abs/1711.10684 and https://arxiv.org/abs/1911.0706, respectively. In the case of the ResUNet++, left the attention block commented out, as I found it to hamper performance in my experiments.

### CODE REVIEW

**Comment by JulienMaille** [comment]
I don't wanna talk for @qubvel but he's gonna ask that you add information about Encoder in following places:

- in README.md: modify number of architectures in start description
- in README.md: add architectures names to table
- in docs/models.rst: add information about model

**Comment by jlcsilva** [comment]
Added the missing documentation. Hope it's ok now. Thanks for your comments!
In the meanwhile, did some experiments in which bilinear interpolation seems to be working better for upsampling. Maybe the kind of upsampling (nearest, bilinear or deconv) could be a possible argument for the UNet family models, but will leave that for a separate pull request.

**Comment by JulienMaille** [comment]
I'm curious what benchmark you have Unet vs ResUnet++ on your own datasets.

**Comment by jlcsilva** [comment]
In the dataset I'm currently working with the UNet actually performs better. I'm working with coronary angiography images with 3 classes (background, artery and catheter) and so far I have obtained the following Dice Scores and Generalized Dice Scores (see https://arxiv.org/abs/1707.03237). 

UNet
- Background: 0.9916 +/- 0.0035
- Catheter: 0.4870 +/- 0.1681
- Artery: 0.7738 +/- 0.0671
- Generalized Dice Score: 0.8250 +/- 0.0677

UNet++
- Background: 0.9930 +/- 0.0031
- Catheter: 0.6105 +/- 0.2039
- Artery: 0.8076 +/- 0.0463
- Generalized Dice Score: 0.8531 +/- 0.0429 

ResUNet:
- Background: 0.9916 +/- 0.0034
- Catheter: 0.5156 +/- 0.1654
- Artery: 0.7768 +/- 0.0508
- Generalized Dice Score: 0.8285 +/- 0,0498

ResUNet++
- Background: 0.9906 +/- 0.0036
- Catheter: 0.4599 +/- 0.1535
- Artery: 0.7670 +/- 0.0502
- Generalized Dice Score: 0.8230 +/- 0.0482

However, I'd say my dataset is not the best to draw conclusions from. Currently I only have 40 samples (30 for training and 10 for testing) and the task is somewhat different from most segmentation tasks (sample in attachment).
783_-12 3_-22_38_IM-0150-0075
783_-12 3_-22_38_IM-0150-0075a

**Comment by JulienMaille** [comment]
I don't wanna hijack your PR, but why do you consider the background as a class and not just have catheter and artery?
I never used generalized dice score, do you also have a generalized dice loss?
Regarding suggestion, I can't pretend to give any. I worked with small datasets too and noticed that reducing the encoder channels did help.

**Comment by jlcsilva** [comment]
For me it's ok, be my guest to hijack it xD Just hope we're not annoying everyone else. Honestly, I don't have a strong reason for it. I'm quite new to segmentation and deep learning in general and it seemed to be a plausible approach. Would you suggest considering only 2 classes and predicting a pixel to be background in case both probabilities were below 0.5, for example? Do you think it would improve performance?
Yeah, it works in the same way as dice score and dice loss. Generalized dice loss = 1 - generalized dice score. Combined with weighted cross entropy, it has proven to be extremely helpful to mitigate the effects of class imbalance in my dataset. Without the weights in the dice loss and the cross entropy the models were simply ignoring the catheters, due to their small area. If you're interested in other loss functions, you might find [this](https://arxiv.org/abs/1707.00478) helpful too. The paper explains it better, but it can be seen as a way to model semantic between classes (objects of some classes are naturally physically closer to objects of certain other classes, and therefore mistaking some of their pixels' classes is more common). Actually, I wanted to test this loss, but still haven't had the time to implement it. Will probably try to do so and make a pull request for it in the near future.

**Comment by JulienMaille** [comment]
I, too, am very new to deep-learning. I work on a 2 class (+background) problem where most of the time, the labels cover less than 2% of the picture. I also have pictures without any class or just one, which raised questions like this one: https://github.com/qubvel/segmentation_models.pytorch/discussions/305
I'm currently using jaccard-loss and micro-IoU but like you, without strong reason for it. I find it difficult to find places (forums, discord, etc.) where newbies like us can discuss and share their trials and errors.

**Comment by JulienMaille** [comment]
@jlcsilva Have you had time to test segmentation without any background class?

**Comment by jlcsilva** [comment]
> @jlcsilva Have you had time to test segmentation without any background class?

Sorry for the late reply. Yeah, I did, but with poor results. Actually, even in binary segmentation, I've gotten better results when explicitly considering both the foreground and background classes.

For me, generalized dice loss (GDL) and data augmentation were the key to achieve reasonable performance in less common classes. [Focal loss](https://arxiv.org/abs/1708.02002), despite having been designed for object detection and not segmentation, also helped in problems were precise boundaries were needed. However, most performance gains come from GDL. My GDL code is kinda messy, but I can share it if you want

**Comment by jlcsilva** [comment]
Closing this because I've updated the code and made a separate pull request containing these and another decoder architecture
